types/account.go

```diff
@@ -13,251 +13,422 @@ import (
+
+// ConsAddress defines a wrapper around bytes meant to present a consensus node.
+// When marshaled to a string or JSON, it uses Bech32.
+type ConsAddress []byte
+
+// ConsAddressFromHex creates a ConsAddress from a hex string.
+func ConsAddressFromHex(address string) (addr ConsAddress, err error) {
+	if len(address) == 0 {
+		return addr, errors.New("decoding Bech32 address failed: must provide an address")
+	}
+
+	bz, err := hex.DecodeString(address)
+	if err != nil {
+		return nil, err
+	}
+
+	return ConsAddress(bz), nil
+}
+
+// ConsAddressFromBech32 creates a ConsAddress from a Bech32 string.
+func ConsAddressFromBech32(address string) (addr ConsAddress, err error) {
+	bz, err := GetFromBech32(address, Bech32PrefixConsAddr)
+	if err != nil {
+		return nil, err
+	}
+
+	return ConsAddress(bz), nil
+}
+
+// Returns boolean for whether two ConsAddress are Equal
+func (ca ConsAddress) Equals(ca2 ConsAddress) bool {
+	if ca.Empty() && ca2.Empty() {
 		return true
 	}
-	return bytes.Compare(bz.Bytes(), bz2.Bytes()) == 0
+
+	return bytes.Compare(ca.Bytes(), ca2.Bytes()) == 0
 }
 
-// Returns boolean for whether an AccAddress is empty
-func (bz ValAddress) Empty() bool {
-	if bz == nil {
+// Returns boolean for whether an ConsAddress is empty
+func (ca ConsAddress) Empty() bool {
+	if ca == nil {
 		return true
 	}
-	bz2 := ValAddress{}
-	return bytes.Compare(bz.Bytes(), bz2.Bytes()) == 0
+
+	ca2 := ConsAddress{}
+	return bytes.Compare(ca.Bytes(), ca2.Bytes()) == 0
+}
+
+// Marshal returns the raw address bytes. It is needed for protobuf
+// compatibility.
+func (ca ConsAddress) Marshal() ([]byte, error) {
+	return ca, nil
+}
+
+// Unmarshal sets the address to the given data. It is needed for protobuf
+// compatibility.
+func (ca *ConsAddress) Unmarshal(data []byte) error {
+	*ca = data
+	return nil
+}
+
+// MarshalJSON marshals to JSON using Bech32.
+func (ca ConsAddress) MarshalJSON() ([]byte, error) {
+	return json.Marshal(ca.String())
+}
+
+// UnmarshalJSON unmarshals from JSON assuming Bech32 encoding.
+func (ca *ConsAddress) UnmarshalJSON(data []byte) error {
+	var s string
+
+	err := json.Unmarshal(data, &s)
+	if err != nil {
+		return nil
+	}
+
+	ca2, err := ConsAddressFromBech32(s)
+	if err != nil {
+		return err
+	}
+
+	*ca = ca2
+	return nil
+}
+
+// Bytes returns the raw address bytes.
+func (ca ConsAddress) Bytes() []byte {
+	return ca
+}
+
+// String implements the Stringer interface.
+func (ca ConsAddress) String() string {
+	bech32Addr, err := bech32.ConvertAndEncode(Bech32PrefixConsAddr, ca.Bytes())
+	if err != nil {
+		panic(err)
+	}
+
+	return bech32Addr
+}
+
+// Format implements the fmt.Formatter interface.
+func (ca ConsAddress) Format(s fmt.State, verb rune) {
+	switch verb {
+	case 's':
+		s.Write([]byte(fmt.Sprintf("%s", ca.String())))
+	case 'p':
+		s.Write([]byte(fmt.Sprintf("%p", ca)))
+	default:
+		s.Write([]byte(fmt.Sprintf("%X", []byte(ca))))
+	}
```

Comment:
small side point - it may be nice to reuse some of the basic logic for the different address types such as this function `Format` - we could create an address interface to implement some of this stuff. Probably for a later refactor

Reply:
I actually had an `Address` interface implemented, but then later decided to remove it as I thought it was too much overhead for the PR.

Reply:
yeah "for a future pr"